# Assignment: 1 - Encoding
### Name: Lilith Bertens
### Hva ID: Bertenl

# Exercise 1A
Encode the following numbers in the basis encoding pattern: 
2, 3, 5, 7, 11, 13, 17, 19, 23.

### Answer
Sure sure, first ill convert it into binary.


In [2]:
array = {2, 3, 5, 7, 11, 13, 17, 19, 23}
Output = ""
for i in array:
    print(bin(i)[2:])

10
11
101
111
1011
1101
10001
10011
10111


Then writing this as $\ket{}$ will give us the number encoded in base.     
So the numbers encoded in a basic manner are:
$\ket{10}$, $\ket{11}$, $\ket{101}$, $\ket{111}$, $\ket{1011}$, $\ket{1101}$, $\ket{10001}$, $\ket{10011}$, $\ket{10111}$    
This is done according to:   
![](img/basis_encoding_solution.png)

# Exercise 1B
How would you encode a negative number?

### Answer
You would use the first bit as a signed bit. If it's one, the number is negative.

# Exercise 1C
Is it possible to encode multiple numbers in one go?

### Answer
Wel that depends, doesn't it? 
if you know and design the circuit with the fact that you're using the first 3 qbits for one number and the last 3 for another and are using an encoder which can convert 2 numbers at one time then sure. but normally? no.

# Exercise 2 
Encode the following numbers using the angle encoding pattern:   
29, 31, 37, 41, 43, 47, 53, 59.

### Answer
Wel, lets create some code which does this for us.    
Im doing this according to:   
![](img/angle_encoding_solution.png)

In [15]:
import qiskit

Array = {29, 31, 37, 41, 43, 47, 53, 59}

for i in Array:
    STR = bin(i)[2:]
    QC = qiskit.QuantumCircuit(STR.__len__())
    print(STR)
    for z in range(STR.__len__()):
        if STR[z] == "0":
            QC.ry(2*0, z)
        else:
            QC.ry(2*1, z)
    print(qiskit.quantum_info.Statevector(QC))


100101
Statevector([0.15772861+0.j, 0.24564775+0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.24564775+0.j,
             0.3825737 +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.24564775+0.j,
             0.3825737 +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.3825737 +0.j, 0.59582324+0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.        +0.j, 0.        +0.j, 0.        +0.j,
             0.  

# Exercise 3
Encode the following numbers using the amplitude encoding pattern:   
61, 67, 71, 73, 79, 83, 89, 97.

### Answer
The code is used to implement this:

In [24]:
import qiskit

Array = {61, 67, 71, 73, 79, 83, 89, 97}

for i in Array:
    STR = bin(i)[2:]
    QC = qiskit.QuantumCircuit(STR.__len__())
    QVEC = qiskit.QuantumCircuit(STR.__len__())
    for z in range(STR.__len__()):
        if STR[z] == "1":
            QVEC.x(z) #im creating the statevector of the circuit first because it refused to cooperate when i just used an int or string.
    QC.prepare_state(qiskit.quantum_info.Statevector(QVEC))
    print(qiskit.quantum_info.Statevector(QC))

Statevector([ 6.12323400e-17+0.j, -7.27081794e-35+0.j,  2.00699663e-35+0.j,
              6.14729979e-51+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
              0.00000000e+00+0.j,  0.00000000e+00+0.j,  0.00000000e+00+0.j,
            

# Exercise 4  
Create a uniform superposition pattern

### Answer
Sure, see the code below:

In [26]:
import qiskit

Qbitnum = 5 # you can change how many qubits you want a uniform superposition for.
QC = qiskit.QuantumCircuit(Qbitnum)
for i in range(Qbitnum):
    QC.h(i)
print(qiskit.quantum_info.Statevector(QC))

Statevector([0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j,
             0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j, 0.1767767+0.j],
            dims=(2, 2, 2, 2, 2))
